MetaMask/metamask-extension
# feat: Change RPC Method Middleware default rate limit to random sampling and add global rate limit

## **Description**

The existing rate limit system in the RPC Method Tracking Middleware was created with the goal of determining which RPC methods were no longer used (i.e. when can we drop eth_sign). This is not random sampling. Currently by default, any methods not explicitly listed are rate limited. This means that for the unlisted methods (i.e. non eth_sign), we also will only record the first usage every 5 minutes. This biases our dataset by under representing the most used methods and over representing the least used.

Our goal is to get broad insights, for which we are not 100% sure what we want to determine yet. This requires random sampling. It may make more sense (and be more accurate!) to use Math.random() as a rate limiter that does not bias by the individual method. This PR achieves that by adding a new random sample rate which is used by default instead of timeout based rate limiting in the RPC method tracking middleware.

We also want to make sure we don't track more events that we do with our current rate limiting strategy. To address that, I have added an additional global rate limit (shared across all methods) that applies after the first rate limit strategy.

Finally, we add `eth_chainId`, `eth_accounts`, and `metamask_logWeb3ShimUsage` to the blocked list which are noisy events that provide us no valuable insights. These 3 events make up about half our tracked rpc method events.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2168

## **Manual testing steps**

Testing rate sampling
1. Opt into MetaMetrics in the wallet settings
1. Open your the background console for extension
1. Add a breakpoint for trackEvent inside `createRPCMethodTrackingMiddleware.js`
1. On any page with a provider injected, use `window.ethereum.request` to make some rpc method calls that are sampled. I suggest spamming `eth_blockNumber`  
1. Check background console for breakpoint trigger. The current sample percentage is 0.1%, so you may have to spam eth_blockNumber for quite some time.

Testing global rate limiting
1. Opt into MetaMetrics in the wallet settings
1. Open your the background console for extension
1. Add a breakpoint for trackEvent inside `createRPCMethodTrackingMiddleware.js`
1. On any page with a provider injected, use `window.ethereum.request` to make some rpc method calls that not rate limited. I suggest using `eth_signTypedData_v4` which is most easily triggerable from the [test dapp](https://metamask.github.io/test-dapp/).
1. Check background console for breakpoint trigger. It should only fire 10 times within 5 minutes of the first request being tracked. After which method calls will not be tracked until additional head room is unlocked as 5 minutes pass since each previous tracked method call.

NOTE: I recommend breakpoint and debugger over the network tab because I don't believe the tracking requests actually fire on dev.

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
